chain/chain/src/store/mod.rs

```diff
@@ -282,6 +283,13 @@ pub struct ChainStore {
     save_trie_changes: bool,
     /// The maximum number of blocks for which a transaction is valid since its creation.
     pub(super) transaction_validity_period: BlockHeightDelta,
+
+    // TODO make caches evict oldest blocks when growing above threashold.
+    // Refcell for inner mutability, has traits require getters to take immutable `&self`.
+    /// Blockheaders are indexed by their hash.
+    block_cache: RefCell<HashMap<CryptoHash, Block>>,
+    block_header_cache: RefCell<HashMap<CryptoHash, BlockHeader>>,
+    block_hash_by_height_cache: RefCell<HashMap<BlockHeight, CryptoHash>>,
```

Comment:
FWIW I think it might make more sense to maintain a view of the most recent blocks in memory (pending analysis into access patterns), rather than caching separate queries and trying to implement something like LRU-evicted cache here.

This would among other things potentially make the cache maintenance somewhat more straightforward and easier to reason about. First, because the only place where the block information is written is in `ChainStoreUpdate::finalize`, so this is where we'd update the cache too. And second is because some stray query (which can happen due to these queries being dependent on the user input) for a block from yesterday ago wouldn't affect the happy path.

Reply:
> rather than caching separate queries

After making `ChainStoreUpdate::finalize` update the cache, I tried to return `cache.block(h).header()` from `get_block_header`. That made a lot of tests fail and turns out some columns get special treatment in `finalize` or garbage collection. So the safest way might be to have a separate cache for each `DBCol`, even if it's a bit of a pain to handle. See the comment on `ChainStoreCache`.

The `apply-range` command doesn't trigger `ChainStoreUpdate::finalize`. I suppose finding a way to do it manually or to otherwise populate the caches for `apply-range` shouldn't be too hard. As a temporary workaround I'm inserting cache-missed data in `get_block, ...` into the cache. More details in the comments.